src/prefect/_internal/concurrency/futures.py

```diff
@@ -0,0 +1,201 @@
+
+
+@dataclasses.dataclass
+class WorkItem:
+    """
+    Transport of a deferred function call.
+    """
+
+    future: concurrent.futures.Future
+    fn: Callable
+    args: Tuple
+    kwargs: Dict[str, Any]
+    context: contextvars.Context
+
+    @classmethod
+    def from_call(cls, __fn, *args, **kwargs) -> "WorkItem":
+        return cls(
+            future=concurrent.futures.Future(),
+            fn=__fn,
+            args=args,
+            kwargs=kwargs,
+            context=contextvars.copy_context(),
+        )
+
+    def __post_init__(self):
+        logger.debug("Created work item %r", self)
+
+    def run(self):
+        """
+        Execute the work item.
+
+        All exceptions during execution of the work item are captured and attached to
+        the future.
+        """
+        # Do not execute if the future is cancelled
+        if not self.future.set_running_or_notify_cancel():
+            return
+
+        logger.debug("Running work item %s", self)
+
+        # Execute the work and set the result on the future
+        if inspect.iscoroutinefunction(self.fn):
+            if get_running_loop():
+                # If an event loop is available, return a coroutine to be awaited
+                return self._run_async()
+            else:
+                # Otherwise, execute the function here
+                return asyncio.run(self._run_async())
+        else:
+            return self._run_sync()
+
+    def _run_sync(self):
+        try:
+            result = self.context.run(self.fn, *self.args, **self.kwargs)
+        except BaseException as exc:
+            self.future.set_exception(exc)
+            # Prevent reference cycle in `exc`
+            self = None
+        else:
+            self.future.set_result(result)
+
+        logger.debug("Finished work item %r", self)
+
+    async def _run_async(self):
+        loop = asyncio.get_running_loop()
+        try:
+            # Call the function in the context; this is not necessary if the function
+            # is a standard cortouine function but if it's a synchronous function that
+            # returns a coroutine we want to ensure the correct context is available
+            coro = self.context.run(self.fn, *self.args, **self.kwargs)
+
+            # Run the coroutine in a new task to run with the correct async context
+            task = self.context.run(loop.create_task, coro)
+            result = await task
+
+        except BaseException as exc:
+            self.future.set_exception(exc)
+            # Prevent reference cycle in `exc`
+            self = None
+        else:
+            self.future.set_result(result)
+
+        logger.debug("Finished work item %r", self)
+
+
+class WatchingFuture(abc.ABC):
+    """
+    A watching future allows work to be sent back to the thread that owns the future by
+    the call the future represents.
+
+    Work sent back to the thread will be executed when the owner waits for the result
+    of the future.
```

Comment:
This docstring is intellectually load-bearing and maybe should be incorporated into a future module doc for this.  This is the clearest explanation for it that we've got so far.  Still think "watching" is the wrong word and I'll keep thinking on it...

Reply:
Yes I agree an overview at the module level would make a lot of sense, perhaps even in `concurrency/__init__.py`